tests/func/test_checkout.py

```diff
@@ -744,3 +744,23 @@ def test_checkout_dir_compat(tmp_dir, dvc):
     remove("data")
     dvc.checkout()
     assert (tmp_dir / "data").read_text() == {"foo": "foo"}
+
+
+def test_checkout_for_files_with_explicit_pull_option_set(tmp_dir, dvc, copy_script):
+    stages = tmp_dir.dvc_gen({"explicit_1": "x", "explicit_2": "y", "always": "z"})
+    for stage in stages[:-1]:
+        stage.outs[0].pull = False
+        stage.dump()
+
+    remove(tmp_dir / "explicit_1")
+    remove(tmp_dir / "explicit_2")
+    remove(tmp_dir / "always")
+
+    dvc.checkout(force=True)
+    assert not (tmp_dir / "explicit_1").exists()
+    assert not (tmp_dir / "explicit_2").exists()
+    assert (tmp_dir / "always").read_text() == "z"
+
+    dvc.checkout(targets="explicit_1.dvc", force=True)
```

Comment:
@dberenbaum @shcheklein, this looks odd to me, that a `pull` option is affecting `checkout`.

I don't think we should support this at all. 

Checkout will fail if we haven't fetched it already as before with the following message:
```
dvc.exceptions.CheckoutError: Checkout failed for following targets:
   explicit_1
   explicit_2
Is your cache up to date?
   <https://error.dvc.org/missing-files>
```


`pull()` should work as expected though, and not fail.

Reply:
@skshetry I understand your concern, what is the practical thing here for users though? any idea? If you have a few targets that have `pull: false` you will keep hitting this failure? do we need an explicit option for checkout then - something like "--allow-missing" in `dvc pull`? 

what is the expected behavior for `dvc pull` itself considering that it is `dvc fetch` + `dvc checkout`?

probably they should be aligned ...

Reply:
That was my intuition as well, that `pull` = `fetch` + `checkout`, so the option should apply to both.

Reply:
Even though it is a combination of `fetch` and `checkout`, it does not necessarily mean `pull` options should apply directly to `fetch` and `checkout` at all. `pull` is a higher-level command.

Reply:
@skshetry do, what is your practical suggestion here?

Reply:
My suggestion would be to: 1) either think of a better option name that captures what this is doing, or, 2) make this apply only to `pull`.

From what I understand in #2825, the feature request is only about `pull`, not `checkout`. And, `checkout` already has `--allow-missing` to help in this scenario.
So, I am more in favor of the second option.

Reply:
If I may add my opinion here, I'd favor the first of your proposals, since the intent here is to provide defaults for `pull`/`fetch`/`checkout`, and requiring the user to supply `--allow-missing` defeats that purpose.

Reply:
okay, a summary of my thoughts:

- w/o CLI options that specifically alter the checkout behavior it should behave the same inside `dvc pull` and when you run `dvc checkout` (since otherwise explanation becomes more complicated, more confusion, etc)
- I also think that checkout should not be affected by the `pull: false` or whatever name we decide. I agree with @skshetry on this I guess. Since it seems reasonable to expect targets to be instantiated if they are already in cache. E.g. I don't want to pull (fetch) / push, but I'd like to still do `dvc checkout`. 
- it is indeed can be annoying for a scenario where users don't want those targets at all on their machine. Clearly they don't want `pull` to fail.

So, how can we reconcile all those objectives here?

Some suggestions:

- rename to `fetch: false` to make sure that this is only about downloading and let people use `--allow-missing` for pull / checkout if they don't want those to fails. It hits all the items, but requires an extra option. If we don't have better options I would prefer this one since it's more explicit, consistent, etc.

- introduce an extra config / option - "treat missing fetch false as warning on checkout"

- make these option ^^ default behavior for the regular checkout and for the `dvc pull`

Any other thoughts?

Reply:
I feel like we're getting hung up on the naming, and letting it drive the implementation. The original intent was to have outputs that are only interacted with via explicit mention on the command line and excluded from all default `fetch`/`checkout`/`pull`. My original implementation named the configuration parameter `explicit`. I understand that was too generic, but maybe we can move back that direction, in a way that doesn't tie the implementation directly to existing subcommands.

Reply:
I think @pbailey-hf has shown a typical use case where it is most likely that the user wants both `fetch` and `checkout` to skip that output.

If we can't agree on a single option that does both, we could add `fetch: false` and `checkout: false` options to accomplish the same, although I don't love needing to set 2 options.

Reply:
Trying to reach some conclusion here - I'm okay with `explicit` (or any other name suggestions anyone has) if it also implies skipping push. This might be more useful than `push: false` since most likely in these cases the user wants to skip push, pull, and checkout for those outputs.

@skshetry @shcheklein Thoughts?

Reply:
>  This might be more useful than push: false since most likely in these cases the user wants to skip push, pull, and checkout for those outputs.

hmm ... this is not obvious to me. A reasonable use case it to keep files locally but avoid pushing them (checkout still works). It feels like an impossible task to create a single option that work the same across all three commands. 

We can it on top of the basic building blocks if we see that it's a convenient and a prevalent way.

> I feel like we're getting hung up on the naming, and letting it drive the implementation. 

Naming-wise, explicit still seems to be a wrong name (explicit what? it needs clarification). But naming is minor. Semantics is more important here.

Sorry @pbailey-hf for all these delays, btw. It just feels important to get it right - otherwise we'll need to maintain this, deprecated, revisit, etc and it's usually painful for everyone involved.

Reply:
@pbailey-hf how important it is for you to avoid checkout if data is already in cache? how bad it is to use allow missing? can we start with that? (see my first suggestion) in one of the comments above?

Reply:
an alternative, re the name: `sync: push | pull | both | all | checkout` - it can be a single option that pretty much encompasses everything

Reply:
> Sorry @pbailey-hf for all these delays, btw. It just feels important to get it right - otherwise we'll need to maintain this, deprecated, revisit, etc and it's usually painful for everyone involved.

No hard feelings. I appreciate the need to get this right, and I appreciate you all doing the maintenance on it.

> @pbailey-hf how important it is for you to avoid checkout if data is already in cache?

That's not as important to me. Mostly I want to avoid large, unnecessary downloads.

> how bad it is to use allow missing? can we start with that? (see my first suggestion) in one of the comments above?

I like this less, since my hope was to have configurable sane defaults for `dvc pull`. Would you be open to implicit "allow missing" behavior for `pull:false` outputs? I suppose we'd need to add that behavior to `dvc push` as well.

Reply:
I took the liberty of implementing the latter. Feedback welcome!

Reply:
> an alternative, re the name: `sync: push | pull | both | all | checkout` - it can be a single option that pretty much encompasses everything

I think this is reasonable and aligns with what I understood `explicit` to be trying to do - skip syncing those outputs unless you _explicitly_ pass them as targets (so `dvc push/pull/checkout` skips it but `dvc push/pull/checkout skipped_target` doesn't). 

I'm not sure we need all of `push | pull | both | all | checkout` now because we already have `push: false` and can't deprecate it right away. We could try a minimal version of `sync: true | false` to start and see with more time which of `push | pull | both | checkout` are needed.

Reply:
@dberenbaum 

> We could try a minimal version of sync: true | false to start

And we would avoid all the operation on it? push, pull, checkout? I think that's fine. I think we can keep granular `push: false` and even probably `fetch: false` (see below). People will be asking about this. `sync` is too broad and won't be convenient in a lot of cases. But I'm fine to start there. @skshetry what is your take?

> Would you be open to implicit "allow missing" behavior for pull:false outputs?

It should be `fetch: false` (not `pull:false` to emphasize that if affects _only_ fetch. And `checkout --allow-missing` behaves the same way as it does now (it should not be trying to analyze lower level fetch flags, etc). This way `pull` and `checkout` are consistent in their behavior - that is important here.

I think we can add a config option for checkout to allow missing by default.

would that combination work for you? 

- `fetch: false`
- config option to imply `allow-missing`?

>

Reply:
@pbailey-hf What do you think about `sync: false`? If you are okay starting there, please go ahead with it, and if you want, you could also add `fetch: false` or other granular options.

Reply:
Okay, thanks for the feedback and your patience @pbailey-hf! I worry about the complexity of so many flags to configure, but let's go with your preference for now to move forward. Please ping when you would like another review.